# <font color=darkred><center>CLAREYAMAR <br>
### <font color=gold><center>Clara Lelis, Yasmin Shimizu, Gabriel Martins<br>
 <font color=darkred><center>*Inluminans Ignotus*☀️🌊
 
<font color=black><center>Divisão responsável pela segurança das águas territoriais e regiões portuárias do reino de Lumi, iluminando a escuridão dos mares e as profundezas do oceano.
    
### <font color=blue><center> Quests 3
<font color=black><center>Após realizar as missões dos murais da Taverna do Sol e da Praça da Fonte e deixar seus relatórios nas mãos da Dra. Diana Didatolov, a Divisão Clareyamar ganhou muito mais reconhecimento e prestígio no reino. Com as missões anteriores concluídas, a Diretora da Biblioteca Real revelou ao grupo mais um mural que contém quests de nível mais elevado , solicitando ajuda a eles. A fim de continuar ajudando o povo e conquistar a Legião da Alvorada, o grupo aceitou e se propôs, novamente, a resolver problemas de diversos cidadãos de Lumi. Além disso, há relatos de que as águas locais estão contaminadas, e sábios marinheiros afirmaram avistar um terrível dragão do mar na região. Conhecendo esses rumores, como divisão responsável pelos recursos hídricos, o grupo decidiu se preparar para estudar o que está acontecendo no Reino de Lumi.
    
##### <font color=darkred><center> Missão 1: As hipóteses reais
<div class="alert alert-warning">
<center> "Vossa Iluminação levantou a hipótese abaixo durante sua última reunião com a cúpula de Ciência, Tecnologia e Encantamentos do reinado. Sua tarefa é respondê-la. Para esta quest, utilize o dataset que pretende usar no seu projeto de conclusão do semestre desta disciplina e escolha o algoritmo de aprendizado de máquina que preferir.
    
<center>- Realizar normalização padrão dos atributos melhora a performance de modelos induzidos pelo algoritmo escolhido."
</div>
    
#### <font color=darkred><center> Introdução da missão 1 - Modelos 4<br>
<font color=black><center> Milu Iluminarius MLVI, o Rei do Reino de Lumi, teve um encontro com Vostrasirus Baba, o vidente mais famoso do reino, na ocasião, o vidente previu a chegada iminente de um terrível dragão à Lumi. O Rei ficou muito preocupado, mas também desconfiado de Vostrasirus, pois lendas contam que dragões haviam sido extintos há muitos séculos. Para testar Vostrasirus, o rei o fez algumas perguntas sobre previsões e seus assistentes registraram as respostas. No intuito de descobrir se as afirmações do guru estão corretas, o rei pediu ajuda no Mural de Quests da Biblioteca Real para que estudantes em previsões analizassem a situação.
    
<font color=black><center>Com isso, a hipótese que será testada nesse caderno é a de que "Realizar normalização padrão dos atributos melhora a performance de modelos induzidos pelo algoritmo escolhido". O teste será realizado comparando as performances dos modelos treinados com os dados originais e com os dados normalizados.

# <font color=gold>Não há normalidade em Lumi

Para realizar as devidas manipulações nos dados, foram utilizadas as bibliotecas `os`,`pandas` e `numpy`.

In [249]:
import os
import pandas as pd
import numpy as np

O dataset usado neste trabalho foi o *Water Quality*, obtido na plataforma *Kaggle*, ele contém informações de corpos d'água como pH, turbidez, condutividade, além de índices de contaminantes como sulfatos, carbonos orgânicos e trialometanos. O arquivo pode ser baixado no site do Kaggle e vem em formato *csv*.

In [250]:
local=os.getcwd()

df = pd.read_csv(local+'\\water_potability.csv')
df = df.dropna()

Para analisar a performance dos modelos, utilizou-se como métrica a Raiz do Erro Quadrático Médio (RMSE).

In [251]:
from sklearn.metrics import mean_squared_error

def calcula_RMSE(y_verdadeiro, y_previsto):
    """Recebe um conjunto de dados conhecidos e um conjunto de dados previstos, compara ambos calculando a raíz do erro quadrático médio e retorna o valor de RMSE obtido."""

    RMSE = mean_squared_error(y_verdadeiro, y_previsto, squared=False)

    return RMSE

Com o dataset definido, é necessário realizar uma divisão nos dados em dados de treino e dados de teste, o que é feito pela função **train_test_split** do `scikit.learn`. Definiu-se que 10% dos dados será de teste e 90% de treino.

In [252]:
from sklearn.model_selection import train_test_split

TAMANHO_TESTE = 0.1
SEMENTE_ALEATORIA = 2110024

indices = df.index
indices_treino, indices_teste = train_test_split(
    indices, test_size=TAMANHO_TESTE, random_state=SEMENTE_ALEATORIA
)

df_treino = df.loc[indices_treino]
df_teste = df.loc[indices_teste]

In [253]:
df_treino

ph    Hardness        Solids  Chloramines     Sulfate  \
52     6.652488  145.010172  19871.788448     4.961066  288.052192   
1059   7.617149  196.606942  27080.268722     7.382755  345.179887   
3209   4.857803  222.399990  15668.273618     7.900790  320.394529   
2211   4.808040  186.160570  19118.050299     7.054444  359.406236   
1331   6.057068  184.373232  14807.268490     5.753405  384.821967   
...         ...         ...           ...          ...         ...   
2595   7.682762  211.663852  12723.342101     6.672341  380.945603   
2496   6.057905  149.122794  23603.501870     6.537028  302.698631   
660    7.543931  192.234541  24227.961139     7.241343  336.151050   
1403   7.721033  203.354788  35800.699273     7.683399  312.919044   
2075  14.000000  235.677458  24008.992040     5.635029  381.097711   

      Conductivity  Organic_carbon  Trihalomethanes  Turbidity  Potability  
52      545.974994       10.942024        71.727414   3.742090           0  
1059    329.016646        8.547283        60.144702   2.391540           0  
3209    417.804063       15.530563        84.228774   3.966571           1  
2211    577.491371        8.304288        73.153089   3.597760           0  
1331    462.006170       14.753055        34.265860   4.371748           0  
...            ...             ...              ...        ...         ...  
2595    360.107441       14.175317        69.949675   5.044306           0  
2496    393.293478       17.615229        54.217939   4.269753           0  
660     369.427699       18.188873        70.588507   3.703191           1  
1403    424.966995       18.092608        96.442129   4.168487           0  
2075    460.745267       13.452538        72.859468   3.170994           0  

[1809 rows x 10 columns]

In [254]:
df_teste

ph    Hardness        Solids  Chloramines     Sulfate  \
1342  8.248703  201.338857  24912.879705     7.877666  389.088844   
452   7.040245  192.955983  20988.429410     6.422359  320.948844   
192   7.362737  185.692181  16268.448606     5.317019  316.918648   
2504  4.851104  146.627237  20227.643783     6.340534  368.941214   
1177  5.484151  196.791251  19378.359067     5.003731  295.591019   
...        ...         ...           ...          ...         ...   
2053  3.846454  211.757205  14686.283390     7.557873  326.912134   
2307  9.808258  220.049574  34132.067979     9.752751  233.870327   
2439  7.895769  168.724047  28723.617627     9.341475  353.976181   
212   5.304143  160.968358  25299.511557     6.286043  300.500689   
1107  6.799376  239.057680   7986.493239    10.365680  419.881175   

      Conductivity  Organic_carbon  Trihalomethanes  Turbidity  Potability  
1342    415.683448       12.842048        51.859151   3.540532           0  
452     381.357350       15.396455        88.298465   3.993410           0  
192     419.258448       16.024246        64.678914   4.408513           0  
2504    346.520310       15.266726        56.989926   3.644107           0  
1177    348.399033       18.713876        76.414153   4.438858           1  
...            ...             ...              ...        ...         ...  
2053    371.269374       21.765170        35.781529   3.143431           0  
2307    367.044379       13.498665        62.954260   4.481261           1  
2439    517.436676       11.992772        79.860552   4.391731           1  
212     497.790391       15.013030        61.115513   4.165724           0  
1107    373.232137       18.828594        43.440644   3.819985           1  

[202 rows x 10 columns]

Inicialmente, retirou-se o atributo *potability* pois ele apresenta apenas valores binários que não influenciam muito no modelo. Assim, definimos os nossos atributos para o treino e o target para a previsão.

In [255]:
atr_treino = df_treino[["ph","Solids","Chloramines","Sulfate","Conductivity","Organic_carbon","Trihalomethanes","Turbidity"]]
target_treino = df_treino["Hardness"]

atr_teste = df_teste[["ph","Solids","Chloramines","Sulfate","Conductivity","Organic_carbon","Trihalomethanes","Turbidity"]]
target_teste = df_teste["Hardness"]

Nem sempre é eficiente usar todos os atributos possíveis para treinar um modelo, por isso, fizemos uma seleção de atributos pelo método de Remoção Recursiva de Atributos (RFE), desejando obter apenas as 4 features mais relevantes para realizar a previsão. O modelo de aprendizado de máquina que foi utilizado é a **Regressão Linear Lasso**, que segue o mesmo princípio da regressão linear por minímos quadrados, mais adiciona algumas camadas matemáticas com o intuito de melhorar a performance do modelo.

In [256]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import Lasso

NUM_ATRIBUTOS_DESEJADOS = 4

modelo = Lasso()

seletor = RFE(modelo, n_features_to_select=NUM_ATRIBUTOS_DESEJADOS)
seletor.fit(atr_treino, target_treino)

atr_treino_modificado = seletor.transform(atr_treino)
atr_teste_modificado = seletor.transform(atr_teste)

print(f"Número de atributos selecionados: {seletor.n_features_}")
print(f"Colunas selecionadas: {seletor.support_}")
print(f"Ordenação (rank) dos atributos {seletor.ranking_}")

Número de atributos selecionados: 4
Colunas selecionadas: [ True False False  True False  True  True False]
Ordenação (rank) dos atributos [1 3 5 1 2 1 1 4]


Após a seleção, definiu-se que os atributos mais relevantes são: **pH, Sulfate, Organic_carbon e Trihalomethanes**, logo, será apenas com eles que treinaremos nosso modelo. No modelo Lasso, um dos hiperparâmetros é valor de alpha, um coeficiente que será usado para melhorar a aproximação dos coeficientes. Para encontrar o melhor valor de alpha, fizemos uma validação cruzada variando os valores.

In [257]:
from sklearn.linear_model import LassoCV

#implementação do modelo
modelo_lassocv = LassoCV(cv=10)

#fitando o modelo
modelo_lassocv.fit(atr_treino_modificado, target_treino)

#guandando o melhor alpha encontrado
melhor_alpha = modelo_lassocv.alpha_
list_alphas = modelo_lassocv.alphas_

print(f"Os valores de alpha testado foram: {list_alphas}.")
print(f"Dentre os valores testados, o que apresentou melhor desempenho foi {melhor_alpha}.")

Os valores de alpha testado foram: [150.91029447 140.73944571 131.2540781  122.40799253 114.15810352
 106.46422942  99.28889668  92.59715735  86.35641886  80.53628525
  75.10840917  70.04635377  65.32546396  60.92274633  56.81675713
  52.98749786  49.41631784  46.08582341  42.97979316  40.08309895
  37.38163223  34.86223533  32.51263735  30.32139441  28.27783391
  26.37200254  24.59461784  22.93702291  21.39114433  19.9494528
  18.60492645  17.35101667  16.18161622  15.09102945  14.07394457
  13.12540781  12.24079925  11.41581035  10.64642294   9.92888967
   9.25971573   8.63564189   8.05362852   7.51084092   7.00463538
   6.5325464    6.09227463   5.68167571   5.29874979   4.94163178
   4.60858234   4.29797932   4.00830989   3.73816322   3.48622353
   3.25126373   3.03213944   2.82778339   2.63720025   2.45946178
   2.29370229   2.13911443   1.99494528   1.86049264   1.73510167
   1.61816162   1.50910294   1.40739446   1.31254078   1.22407993
   1.14158104   1.06464229   0.99288897   

Após toda essa preparação, conseguimos finalmente aplicar o modelo do modo como queremos, usando o melhor valor de alpha que encontramos e os 3 atributos mais relevantes. Faremos também a avaliação da performance do modelo fazendo o RMSE considerando os dados previstos e os dados de teste. 

In [258]:
from sklearn.linear_model import Lasso

modelo_lasso = Lasso(alpha=melhor_alpha)

modelo_lasso.fit(atr_treino_modificado,target_treino)

previsoes_lasso = modelo_lasso.predict(atr_teste_modificado)

RMSE_lasso = calcula_RMSE(target_teste, previsoes_lasso)

print(f"O RMSE do modelo linear Lasso foi de {RMSE_lasso} mg/L.")

O RMSE do modelo linear Lasso foi de 34.24280909719526 mg/L.


Obtivemos esse valor de RMSE, o que considerando a dimensão dos dados, já parece grande, mas para termos melhor certeza da performance, comparamos com a performance do modelo **baseline**

In [259]:
from sklearn.dummy import DummyRegressor
 
modelo_bl = DummyRegressor()
 
modelo_bl.fit(atr_treino_modificado,target_treino)
 
previsoes_bl = modelo_bl.predict(atr_teste_modificado)
 
RMSE_bl = calcula_RMSE(target_teste, previsoes_bl)
 
print(f"O RMSE do modelo baseline foi de {RMSE_bl} mg/L.")

O RMSE do modelo baseline foi de 34.53543651088391 mg/L.


O RMSE do Baseline foi bastante parecido com o do modelo Lasso, o que nos indica que nosso modelo não é muito bom. A hipótese que foi levantada é a de que isso se deve a má qualidade dos dados do dataset, uma vez que eles apresentam baixíssima correlação, além de informações bastante estranhas e não-condizentes com a realidade. Decidimos, portanto, testar o modelo com o dataset **Penguins**, fornecido pela biblioteca `Seaborn`, afim de obter resultados mais conclusivos.

## <font color=blue>Penguins para nos salvar

O passo a passo da preparação e aplicação do modelo no dataset *penguins* não difere muito da feita com o dataset *Water Quality*. Logo, vamos nos focar nos resultados finais do modelo.

In [260]:
import seaborn as sb

df_p = sb.load_dataset('penguins')
df_p = df_p.dropna()
df_p

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
4    Adelie  Torgersen            36.7           19.3              193.0   
5    Adelie  Torgersen            39.3           20.6              190.0   
..      ...        ...             ...            ...                ...   
338  Gentoo     Biscoe            47.2           13.7              214.0   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    Male  
1         3800.0  Female  
2         3250.0  Female  
4         3450.0  Female  
5         3650.0    Male  
..           ...     ...  
338       4925.0  Female  
340       4850.0  Female  
341       5750.0    Male  
342       5200.0  Female  
343       5400.0    Male  

[333 rows x 7 columns]

In [261]:
indices_p = df_p.index
indices_p_treino, indices_p_teste = train_test_split(
    indices_p, test_size=TAMANHO_TESTE, random_state=SEMENTE_ALEATORIA
)

df_p_treino = df_p.loc[indices_p_treino]
df_p_teste = df_p.loc[indices_p_teste]

Usamos apenas os atributos numéricos para treinar o modelo, uma vez que misturá-los envolveria conhecimentos que não temos ainda.

In [262]:
atr_treino_p = df_p_treino[["bill_length_mm","bill_depth_mm","flipper_length_mm"]]
target_treino_p = df_p_treino["body_mass_g"]

atr_teste_p = df_p_teste[["bill_length_mm","bill_depth_mm","flipper_length_mm"]]
target_teste_p = df_p_teste["body_mass_g"]

In [263]:
modelo_lassocv_p = LassoCV(cv=10)

#fitando o modelo
modelo_lassocv_p.fit(atr_treino_p, target_treino_p)

#guandando o melhor alpha encontrado
melhor_alpha_p = modelo_lassocv.alpha_
list_alphas = modelo_lassocv.alphas_

print(f"Os valores de alpha testado foram: {list_alphas}.")
print(f"Dentre os valores testados, o que apresentou melhor desempenho foi {melhor_alpha_p}.")

Os valores de alpha testado foram: [150.91029447 140.73944571 131.2540781  122.40799253 114.15810352
 106.46422942  99.28889668  92.59715735  86.35641886  80.53628525
  75.10840917  70.04635377  65.32546396  60.92274633  56.81675713
  52.98749786  49.41631784  46.08582341  42.97979316  40.08309895
  37.38163223  34.86223533  32.51263735  30.32139441  28.27783391
  26.37200254  24.59461784  22.93702291  21.39114433  19.9494528
  18.60492645  17.35101667  16.18161622  15.09102945  14.07394457
  13.12540781  12.24079925  11.41581035  10.64642294   9.92888967
   9.25971573   8.63564189   8.05362852   7.51084092   7.00463538
   6.5325464    6.09227463   5.68167571   5.29874979   4.94163178
   4.60858234   4.29797932   4.00830989   3.73816322   3.48622353
   3.25126373   3.03213944   2.82778339   2.63720025   2.45946178
   2.29370229   2.13911443   1.99494528   1.86049264   1.73510167
   1.61816162   1.50910294   1.40739446   1.31254078   1.22407993
   1.14158104   1.06464229   0.99288897   

In [264]:
modelo_lasso_p = Lasso(alpha=melhor_alpha_p)

modelo_lasso_p.fit(atr_treino_p,target_treino_p)

previsoes_lasso_p = modelo_lasso_p.predict(atr_teste_p)

RMSE_lasso_p = calcula_RMSE(target_teste_p, previsoes_lasso_p)

print(f"O RMSE do modelo linear Lasso foi de {RMSE_lasso_p} g.")

O RMSE do modelo linear Lasso foi de 355.2861858684008 g.


In [265]:
from sklearn.dummy import DummyRegressor
 
modelo_bl_p = DummyRegressor()
 
modelo_bl_p.fit(atr_treino_p,target_treino_p)
 
previsoes_bl_p = modelo_bl_p.predict(atr_teste_p)
 
RMSE_bl_p = calcula_RMSE(target_teste_p, previsoes_bl_p)
 
print(f"O RMSE do modelo baseline foi de {RMSE_bl_p} g.")

O RMSE do modelo baseline foi de 774.42427411348 g.


Após isso, obtemos melhores resultados para fundamentarem nossa análise acerca da questão proposta. O modelo Lasso desempenhou um bom resultado, tendo um RMSE de menos da metade do RMSE do baseline. Com esse resultado, vamos agora para a aplicação da normalização dos dados para verificar se faz diferença os dados serem normalizados antes do treino.

# <font color=gold>Normalizando a estranheza

Um dos principais processos em pré-processamento de dados para aplicação de modelos de aprendizado de máquina é a normalização dos dados, que consiste em diminuir a diferença de escala dos dados para que valores muito altos ou muito pequenos não se sobreponham aos demais nas diversas operações matemáticas dos modelos. Um tipo de normalização muito usada é a **Normalização Padrão**, que consiste em transformar a média dos dados em 0 e o desvio padrão em 1. Ela foi aplicada ao dataset *Water Quality* (não vamos desistir dele ainda) usando a função **StandardScaler** da biblioteca `scikit.learn`.

In [266]:
from sklearn.preprocessing import StandardScaler

# Instanciando o normalizador
normalizador = StandardScaler()

# Ajustando o normalizador aos dados
normalizador.fit(df)

# Armazenando os resultados do normalizador
normaldata = normalizador.transform(df)

Por padrão, o que o normalizador devolve é um array dos dados, mas para a aplicação precisamos deles em um **DataFrame**, logo, armazenamos abaixo os dados normalizados em um DataFrame.

In [267]:
df_norm = pd.DataFrame(normaldata)

print(df_norm)

             0         1         2         3         4         5         6  \
0     0.782466  0.564114  0.011687  0.583804  0.574378 -0.783962  1.227032   
1     1.275463 -0.455653 -0.455835 -0.370947 -0.560480 -0.348429 -0.842154   
2    -0.954835 -0.234614  0.790645  0.259104 -0.158911 -1.810063 -1.792340   
3     1.994902  1.596951  0.790764  0.239248  1.467140 -1.770608 -0.170876   
4     0.985323  0.226606 -0.954313 -1.622878 -0.726179  0.595858 -0.599824   
...        ...       ...       ...       ...       ...       ...       ...   
2006  1.210411  0.584770 -0.693978 -0.528284 -0.492625 -0.447578 -1.341281   
2007 -0.243774  0.347964 -0.540564  0.362137 -0.697038 -1.205321  0.559422   
2008  2.800492 -3.100365  1.767503  1.343596 -1.803476  0.165656  0.546021   
2009 -0.646160 -0.285317  0.488576  0.387023  0.302843 -0.131852 -0.688929   
2010 -1.537172 -0.070075  2.970287  0.020386  0.648718  1.238006 -0.139372   

             7         8         9  
0     2.111652  0.844761 -

Ao normalizar o dataframe, altera-se o nome das colunas, transformando-as em números começando de 0, a ordem porém é mantida, o que nos permite entender a quais colunas aqueles dados se referem.

Assim como nos modelos já aplicados acima, realizamos o split de treino e teste.

In [268]:
from sklearn.model_selection import train_test_split

TAMANHO_TESTE = 0.1
SEMENTE_ALEATORIA = 2110024

indices = df_norm.index
indices_treino, indices_teste = train_test_split(
    indices, test_size=TAMANHO_TESTE, random_state=SEMENTE_ALEATORIA
)

df_norm_treino = df_norm.loc[indices_treino]
df_norm_teste = df_norm.loc[indices_teste]

print(df_norm_treino.columns)

RangeIndex(start=0, stop=10, step=1)


Após isso, dividimos novamente entre atributos e target.

In [269]:
atr_norm_treino = df_norm_treino[[0,2,3,4,5,6,7,8]]
target_norm_treino = df_norm_treino[1]

atr_norm_teste = df_norm_teste[[0,2,3,4,5,6,7,8]]
target_norm_teste = df_norm_teste[1]

print('Os atributos normalizados de treino são:')
print(atr_norm_treino)

Os atributos normalizados de treino são:
             0         2         3         4         5         6         7  \
28   -0.275599 -0.236763 -1.371646 -1.096554  1.480293 -1.027542  0.331395   
661   0.337684  0.597543  0.156786  0.290211 -1.208412 -1.747953 -0.389231   
1969 -1.416570 -0.723276  0.483741 -0.311450 -0.108094  0.352830  1.109176   
1359 -1.448207 -0.324000 -0.050425  0.635553  1.870867 -1.821053  0.420095   
818  -0.654137 -0.822928 -0.871567  1.252516  0.439691  0.118932 -1.999302   
...        ...       ...       ...       ...       ...       ...       ...   
1598  0.379398 -1.064121 -0.291587  1.158418 -0.823112 -0.054869  0.220792   
1535 -0.653605  0.195144 -0.376988 -0.741015 -0.411847  0.979961 -0.757969   
406   0.291136  0.267418  0.067536  0.071037 -0.707609  1.152530  0.260537   
862   0.403728  1.606842  0.346536 -0.492916 -0.019325  1.123571  1.869039   
1273  4.395582  0.242075 -0.946279  1.162110  0.424065 -0.272303  0.401827   

             8  
28   

Da mesma maneira, realizamos a seleção de atributos por RFE e a validação cruzada para encontrar o melhor alpha.

In [270]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import Lasso

NUM_ATRIBUTOS_DESEJADOS = 4

modelo = Lasso()

seletor = RFE(modelo, n_features_to_select=NUM_ATRIBUTOS_DESEJADOS)
seletor.fit(atr_norm_treino, target_norm_treino)

atr_norm_treino_modificado = seletor.transform(atr_norm_treino)
atr_norm_teste_modificado = seletor.transform(atr_norm_teste)

print(f"Número de atributos selecionados: {seletor.n_features_}")
print(f"Colunas selecionadas: {seletor.support_}")
print(f"Ordenação (rank) dos atributos {seletor.ranking_}")

Número de atributos selecionados: 4
Colunas selecionadas: [False False False False  True  True  True  True]
Ordenação (rank) dos atributos [5 4 3 2 1 1 1 1]


In [271]:
from sklearn.linear_model import LassoCV

#implementação do modelo
modelo_lassocv_n = LassoCV(cv=10)

#fitando o modelo
modelo_lassocv_n.fit(atr_norm_treino_modificado, target_norm_treino)

#guandando o melhor alpha encontrado
melhor_alpha_n = modelo_lassocv_n.alpha_
list_alphas = modelo_lassocv_n.alphas_

print(f"Os valores de alpha testado foram: {list_alphas}.")
print(f"Dentre os valores testados, o que apresentou melhor desempenho foi {melhor_alpha}.")

Os valores de alpha testado foram: [2.87931663e-02 2.68526032e-02 2.50428277e-02 2.33550249e-02
 2.17809744e-02 2.03130096e-02 1.89439807e-02 1.76672198e-02
 1.64765084e-02 1.53660468e-02 1.43304267e-02 1.33646039e-02
 1.24638743e-02 1.16238509e-02 1.08404423e-02 1.01098327e-02
 9.42846385e-03 8.79301694e-03 8.20039703e-03 7.64771772e-03
 7.13228714e-03 6.65159486e-03 6.20329963e-03 5.78521799e-03
 5.39531366e-03 5.03168758e-03 4.69256868e-03 4.37630525e-03
 4.08135693e-03 3.80628713e-03 3.54975612e-03 3.31051443e-03
 3.08739684e-03 2.87931663e-03 2.68526032e-03 2.50428277e-03
 2.33550249e-03 2.17809744e-03 2.03130096e-03 1.89439807e-03
 1.76672198e-03 1.64765084e-03 1.53660468e-03 1.43304267e-03
 1.33646039e-03 1.24638743e-03 1.16238509e-03 1.08404423e-03
 1.01098327e-03 9.42846385e-04 8.79301694e-04 8.20039703e-04
 7.64771772e-04 7.13228714e-04 6.65159486e-04 6.20329963e-04
 5.78521799e-04 5.39531366e-04 5.03168758e-04 4.69256868e-04
 4.37630525e-04 4.08135693e-04 3.80628713e-04 3.54

Depois dos processos acima, foi possível aplicar o modelo com os atributos e hiperparâmtros definidos.

In [284]:
from sklearn.linear_model import Lasso

modelo_lasso_n = Lasso(alpha=melhor_alpha_n)

modelo_lasso_n.fit(atr_norm_treino_modificado,target_norm_treino)

df_atr_teste_modificado = pd.DataFrame(atr_norm_teste_modificado)

previsoes_lasso = modelo_lasso_n.predict(df_atr_teste_modificado)

RMSE_lasso = calcula_RMSE(target_norm_teste, previsoes_lasso)

print(f"O RMSE do modelo linear Lasso foi de {RMSE_lasso} mg/L.")

O RMSE do modelo linear Lasso foi de 1.0584935461942098 mg/L.


Assim como antes, esse valor de RMSE sozinho não nos diz muito, portanto, comparamos com o desempenho do baseline.

In [286]:
from sklearn.dummy import DummyRegressor
 
modelo_bl = DummyRegressor()
 
modelo_bl.fit(atr_norm_treino_modificado,target_norm_treino)
 
previsoes_bl = modelo_bl.predict(atr_norm_teste_modificado)
 
RMSE_bl = calcula_RMSE(target_norm_teste, previsoes_bl)
 
print(f"O RMSE do modelo baseline foi de {RMSE_bl} mg/L.")

O RMSE do modelo baseline foi de 1.0584935461942098 mg/L.


Os RMSE deram exatamente o mesmo valor até a 16° casa decimal, o que é estranhíssimo, e indica que nosso modelo não é bom, provavelmente devido a **péssima** qualidade do dataset. Vamos, então, voltar para nossos fofos e queridos penguins que têm apresentado melhores resultados.

## <font color=blue>Super Penguins - O Retorno

Aplicaremos o modelo Lasso novamente no dataset *penguins*, mas dessa vez, normalizaremos os dados para treinar o modelo. O processo de normalização segue da mesma maneira como já foi desenvolvido anteriormente.

In [274]:
df_pen = df_p[["bill_length_mm","bill_depth_mm","flipper_length_mm","body_mass_g"]]

In [275]:
normalizador = StandardScaler()

normalizador.fit(df_pen)

normdata_p = normalizador.transform(df_pen)

df_pen_norm = pd.DataFrame(normdata_p,columns=["bill_length_mm","bill_depth_mm","flipper_length_mm","body_mass_g"])
df_pen_norm

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g
0         -0.896042       0.780732          -1.426752    -0.568475
1         -0.822788       0.119584          -1.069474    -0.506286
2         -0.676280       0.424729          -0.426373    -1.190361
3         -1.335566       1.085877          -0.569284    -0.941606
4         -0.859415       1.747026          -0.783651    -0.692852
..              ...            ...                ...          ...
328        0.587352      -1.762145           0.931283     0.892957
329        0.514098      -1.457000           1.002739     0.799674
330        1.173384      -0.744994           1.502928     1.919069
331        0.221082      -1.202712           0.788372     1.234995
332        1.081817      -0.541564           0.859828     1.483749

[333 rows x 4 columns]

- ### <font color=darkred>Split de Treino e Teste

In [276]:
indices_p_norm = df_pen_norm.index
indices_treino_p_norm, indices_teste_p_norm = train_test_split(
    indices_p_norm, test_size=TAMANHO_TESTE, random_state=SEMENTE_ALEATORIA
)

df_p_norm_treino = df_pen_norm.loc[indices_treino_p_norm]
df_p_norm_teste = df_pen_norm.loc[indices_teste_p_norm]

- ### <font color=darkred>Separação dos Atributos e Target

In [277]:
atr_norm_treino_p = df_p_norm_treino[["bill_length_mm","bill_depth_mm","flipper_length_mm"]]
target_norm_treino_p = df_p_norm_treino["body_mass_g"]

atr_norm_teste_p = df_p_norm_teste[["bill_length_mm","bill_depth_mm","flipper_length_mm"]]
target_norm_teste_p = df_p_norm_teste["body_mass_g"]

- ### <font color=darkred>Validação Cruzada para o melhor alpha

In [278]:
modelo_lassocv_p_n = LassoCV(cv=10)

modelo_lassocv_p_n.fit(atr_norm_treino_p, target_norm_treino_p)

melhor_alpha_p_n = modelo_lassocv_n.alpha_
list_alphas = modelo_lassocv_p_n.alphas_

print(f"Os valores de alpha testado foram: {list_alphas}.")
print(f"Dentre os valores testados, o que apresentou melhor desempenho foi {melhor_alpha}.")

Os valores de alpha testado foram: [0.87024074 0.81158943 0.75689102 0.7058791  0.65830521 0.61393764
 0.5725603  0.53397165 0.49798375 0.46442131 0.43312087 0.40392997
 0.37670644 0.35131769 0.32764005 0.30555821 0.28496461 0.26575895
 0.24784768 0.23114358 0.21556528 0.2010369  0.18748768 0.17485164
 0.16306723 0.15207704 0.14182756 0.13226885 0.12335438 0.1150407
 0.10728735 0.10005654 0.09331306 0.08702407 0.08115894 0.0756891
 0.07058791 0.06583052 0.06139376 0.05725603 0.05339716 0.04979837
 0.04644213 0.04331209 0.040393   0.03767064 0.03513177 0.03276401
 0.03055582 0.02849646 0.02657589 0.02478477 0.02311436 0.02155653
 0.02010369 0.01874877 0.01748516 0.01630672 0.0152077  0.01418276
 0.01322689 0.01233544 0.01150407 0.01072873 0.01000565 0.00933131
 0.00870241 0.00811589 0.00756891 0.00705879 0.00658305 0.00613938
 0.0057256  0.00533972 0.00497984 0.00464421 0.00433121 0.0040393
 0.00376706 0.00351318 0.0032764  0.00305558 0.00284965 0.00265759
 0.00247848 0.00231144 0.00215

- ### <font color=darkred>Implementação do modelo

In [279]:
modelo_lasso_p_n = Lasso(alpha=melhor_alpha_p_n)

modelo_lasso_p_n.fit(atr_norm_treino_p,target_norm_treino_p)

previsoes_lasso_p_n = modelo_lasso_p_n.predict(atr_norm_teste_p)

RMSE_lasso_p_n = calcula_RMSE(target_norm_teste_p, previsoes_lasso_p_n)

print(f"O RMSE do modelo linear Lasso foi de {RMSE_lasso_p_n} g.")

O RMSE do modelo linear Lasso foi de 0.42636805605896694 g.


- ### <font color=darkred>Baseline para comparação

In [280]:
from sklearn.dummy import DummyRegressor
 
modelo_bl = DummyRegressor()
 
modelo_bl.fit(atr_norm_treino_p,target_norm_treino_p)
 
previsoes_bl_p_n = modelo_bl.predict(atr_norm_teste_p)
 
RMSE_bl_p_n = calcula_RMSE(target_norm_teste_p, previsoes_bl_p_n)
 
print(f"O RMSE do modelo baseline foi de {RMSE_bl_p_n} mg/L.")

O RMSE do modelo baseline foi de 0.9632072507031078 mg/L.


Finalmente, com os resultados obtidos podemos fazer a devida análise da diferença de desempenho entre modelos treinados com dados normalizados e não-normalizados. Para isso, vamos avaliar os modelos através da razão entre a performance do modelo Lasso e a performande do baseline. A partir disso, quanto menor a razão melhor a performance.

In [281]:
print('Razão da performance nos dados normalizados')
print(RMSE_lasso_p_n/RMSE_bl_p_n)
print()
print('Razão da performance nos dados não normalizados')
print(RMSE_lasso_p/RMSE_bl_p)

Razão da performance nos dados normalizados
0.44265453332886884

Razão da performance nos dados não normalizados
0.45877459907247053


É possível observar que a performance com os dados normalizados foi melhor do que com os dados não-normalizados, logo a afirmação dada é verdadeira.

# <font color=gold>Análise Final

O que podemos concluir a partir do trabalho desenvolvido, é que a afirmação de Vostrasirus está, de fato, correta, a performance do modelo melhora com dados normalizados. Isso no diz que normalizador os dados é uma etapa importante no processo de indução de modelos preditivos. E que Vostrasirus parece não estar mentindo, porém, mais análises são necessárias...

# <font color=blue>Referências

- CASSAR, DR. ATP-203 1.1 - Tratamento de dados. Microsoft Teams, T03-ATP-203 - Aprendizado de Máquina, Material de Aula.

- CASSAR, DR. ATP-203 2.1 - Aprendizado de máquina, k-NN e métricas. Microsoft Teams, T03-ATP-203 - Aprendizado de Máquina, Material de Aula.

- CASSAR, DR. ATP-203 3.0 - Modelo linear e baseline. Microsoft Teams, T03-ATP-203 - Aprendizado de Máquina, Material de Aula.

- CASSAR, DR. ATP-203 4.0 - Split de dados de treino e teste. Microsoft Teams, T03-ATP-203 - Aprendizado de Máquina, Material de Aula.

- CASSAR, DR. ATP-203 5.0 - Validação cruzada, busca aleatória e busca em grade. Microsoft Teams, T03-ATP-203 - Aprendizado de Máquina, Material de Aula.

- CASSAR, DR. ATP-203 7.1 - Seleção de atributos. Microsoft Teams, T03-ATP-203 - Aprendizado de Máquina, Material de Aula.

- Water Quality. Disponível em: <https://www.kaggle.com/datasets/adityakadiwal/water-potability>. Acesso em: 27 de Set. 2024.

- KIREN, M. Lasso regression: a comprehensive guide. Disponível em: <https://kirenz.github.io/regression/docs/lasso.html>. Acesso em: 27 set. 2024.

- SCIKIT LEARN. LassoCV — scikit-learn 0.22 Documentation. Disponível em: <https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LassoCV.html>. Acesso em: 27 de Out. 2024.